Revising subnet 1 code walkthrough document (#163)
opentensor/developer-docs: docs/subnets/walkthrough-prompting.md

For easier understanding, in this document we will focus on how a **single** subnet validator interacts with **multiple** subnet miners in this subnet. 

<center id="bittensor-img">
<ThemedImage
alt="1-Prompting Walkthrough"
sources={{
    light: useBaseUrl('/img/docs/1-prompting-subnet-walkthrough.svg'),
    dark: useBaseUrl('/img/docs/1-prompting-subnet-walkthrough.svg'),
  }}
style={{width: 600}}
/>
</center>

## Subnet 1 summary

See the below diagram showing a high-level view of how this Prompting subnet works. 

<center>
<ThemedImage
alt="Prompting Subnet 1 Big Picture"
sources={{
    light: useBaseUrl('/img/docs/2-prompting-subnet-high-level.svg'),
    dark: useBaseUrl('/img/docs/2-prompting-subnet-high-level.svg'),
  }}
style={{width: 750}}
/>
</center>

The numbered items in the below description correspond to the numbered sections in the above diagram:

1. The subnet validator sends a **challenge** simultaneously to multiple subnet miners. This step constitutes the **prompt** from the subnet validator to subnet miners. A challenge is a prompt sent to subnet miners in such a way that:

>    - This prompt is in a style and tone that is similar to humans.

Reviewer: This is so that miners become adept at handling ‘fuzzy’ instructions with ambiguity - something that is required in order to excel at understanding user needs

   - This prompt drives the conversation between the subnet validator and the subnet miners in order to reach a pre-defined goal. 

    See the below [Challenge generation](#challenge-generation) section for how a challenge is generated. 

> 2. The subnet miners respond to the subnet validator after performing the challenge **task**. 

Reviewer: Most tasks require the miners to make use of APIs or tools to perform well.

> 3. The subnet validator then **scores** each subnet miner by comparing the subnet miner's response with a locally generated **reference** answer. The subnet validator uses this reference as the ground truth for this step. 

Reviewer: The reference is generated using data from APIs and tools to ensure factual correctness and to be able to provide citations.

4. Finally, the subnet validator **sets the weights** for the subnet miners by sending the weights to the blockchain. In the Bittensor blockchain the Yuma Consensus allocates the rewards to the participating subnet miners and subnet validators. 

:::tip Use of large language models 
In this subnet both the subnet validator and the subnet miners use large language models (LLMs) to create the challenges (subnet validator) and respond to the prompts (subnet miners).
:::

## Key innovations in this subnet

This subnet has developed a few innovative techniques to get to a real human-like conversational AI that actually produces intelligence instead of copying from the internet. See below:

### Achieving human-like conversation

To deliver to a user of this subnet an experience of a human-like conversation:

- Subnet validators perform a roleplay where they take on the persona of **random** human users before they prompt the subnet miners. By doing this, the subnet validators engage the subnet miners in a real, random, human-like conversation throughout the subnet operation.

> - This means that throughout the subnet validation process the subnet miners continually improve their performance as though all validation prompts originated from an external API client query. This enables the subnet miners to become better intelligent AI assistants. 

Reviewer: ```suggestion
- This means that throughout the subnet validation process the subnet miners continually improve their performance as though all validation prompts originated from an external API client query. This enables the subnet miners to become better intelligent AI assistants. 
```
Don’t love the way this reads. Maybe integrate my comments above regarding handling ‘fuzzy’ instructions

:::tip Class HumanAgent
See [class HumanAgent](https://github.com/opentensor/prompting/blob/main/prompting/agent.py#L30).
:::

### Prevent subnet miners from looking up the answers

To prevent the subnet miners from simply looking up the answers on the internet, this subnet incorporates a novel approach&mdash;it introduces fuzziness into the prompt and requires that the subnet miners use semantic intelligence to understand the instructions contained in the prompt.

### Evolve subnet as a mixture of experts (MoE)

> The subnet validator composes a challenge based on whether the task is answering questions, summarizing a given text, debugging code, solve mathematics problems. The motivation behind using multiple tasks is several fold:

Reviewer: ```suggestion
The subnet validator composes a challenge based on whether the task is answering questions, summarizing a given text, debugging code, solve mathematics problems and so on. The motivation behind using multiple tasks is several fold:
```

Reviewer: Let’s keep it open ended. We add new tasks all the time to mimic the distribution of real use cases and benchmark the miner performance across a broad skill set (which is also useful for a routing mechanism)

---
title: "Walkthrough of Subnet 1"
---

import ThemedImage from '@theme/ThemedImage';
import useBaseUrl from '@docusaurus/useBaseUrl';

# Walkthrough of Subnet 1

> In this section we present a high-level walkthrough of the [Subnet 1: Prompting](https://github.com/opentensor/prompting/tree/main). This subnet creates  human-conversation-like AI for a variety of tasks such as:

Reviewer: ```suggestion
In this section we present a high-level walkthrough of the [Subnet 1: Prompting](https://github.com/opentensor/prompting/tree/main). This subnet contains multiple incentive mechanisms which produce an internet-scale conversational intelligence. It is driven by large language models (LLMs) which search the internet and utilize specialized simulator modules to produce factually accurate and mathematically correct responses.
```

For easier understanding, in this document we will focus on how a **single** subnet validator interacts with **multiple** subnet miners in this subnet. 

<center id="bittensor-img">
<ThemedImage
alt="1-Prompting Walkthrough"
sources={{
    light: useBaseUrl('/img/docs/1-prompting-subnet-walkthrough.svg'),
    dark: useBaseUrl('/img/docs/1-prompting-subnet-walkthrough.svg'),
  }}
style={{width: 600}}
/>
</center>

## Subnet 1 summary

See the below diagram showing a high-level view of how this Prompting subnet works. 

<center>
<ThemedImage
alt="Prompting Subnet 1 Big Picture"
sources={{
    light: useBaseUrl('/img/docs/2-prompting-subnet-high-level.svg'),
    dark: useBaseUrl('/img/docs/2-prompting-subnet-high-level.svg'),
  }}
style={{width: 750}}
/>
</center>

The numbered items in the below description correspond to the numbered sections in the above diagram:

> 1. The subnet validator sends a **challenge** simultaneously to multiple subnet miners. This step constitutes the **query** from the subnet validator to subnet miners. A challenge is a prompt sent to subnet miners in such a way that:

Reviewer: Feels like we're sweeping quite a lot under the rug with this high level. My opinion is we could dip our toes a little further into the technical (or at least novel) aspects of the subnet. 

For instance, to prevent miners from simply 'looking up' the answers we have developed a novel approach which introduces 'fuzziness' and requires the use of semantic intelligence to understand instructions.

> 1. The subnet validator sends a **challenge** simultaneously to multiple subnet miners. This step constitutes the **query** from the subnet validator to subnet miners. A challenge is a prompt sent to subnet miners in such a way that:

Reviewer: ```suggestion
1. The subnet validator sends a **challenge** simultaneously to multiple subnet miners. This step constitutes the **prompt** from the subnet validator to subnet miners. A challenge is a prompt sent to subnet miners in such a way that:
```
If we can avoid reusing the term query it would be preferred. If you are using this term globally it's okay.

>    - This prompt is in a style and tone that is similar to humans.

Reviewer: I think prompt is the right word choice instead of query

   - This prompt drives the conversation between the subnet validator and the subnet miners in order to reach a pre-defined goal. 

    See below section for how a challenge is generated. 
2. The subnet miners respond to the subnet validator after performing the challenge **task**. 
3. The subnet validator then **scores** each subnet miner by comparing the subnet miner's response with a locally generated **reference** answer. The subnet validator uses this reference as the ground truth for this step. 
4. Finally, the subnet validator **sets the weights** for the subnet miners by sending the weights to the blockchain. In the Bittensor blockchain the Yuma Consensus allocates the rewards to the participating subnet miners and subnet validators. 

:::tip Use of large language models 
In this subnet both the subnet validator and the subnet miners use large language models (LLMs) to create the challenges (subnet validator) and respond to the prompts (subnet miners).
:::

## Challenge generation

> The subnet validator composes a challenge based on whether the task is answering questions, summarizing a given text, debugging code, solve mathematics problems. The challenge-creation works like this:

Reviewer: The motivation behind using multiple tasks is severalfold:
1. It constantly benchmarks the capabilities of the miners across a broad range of challenging tasks that are common use-cases. This can be used to route queries to specialized miners, thereby providing an effective mixture of experts system.
2. It is also a precursor to an inter-subnet bridging mechanism that will enable SN1 to 'reach out' into other subnets are access the useful work that can be provided by them. Miners must become adept at using tools and APIs in order to fulfill validation tasks. We are building an API layer for inter-subnet communication, which is a natural extension of 'agentic' models.
